### PR TITLE
[Feature:Forum] Display pronouns in forum option

### PR DESCRIPTION
**Pull request by yanliw123**
display pronouns option in profile page and function to display pronoun in forum works good, read for review

### Please check if the PR fulfills #9154

* [ ] Tests for the changes have been added/updated (if possible)
display pronouns option in profile page 
My Profile - Google Chrome 2023_6_14 14_14_36

### What is the new behavior?
once checked the box for display pronouns, you should able to see your pronouns next to your name in discussion forum.

### CODE REVIEW

**Review by NiaHeermance — CHANGES_REQUESTED** [review]
I changed Alyssa P. Hacker's pronoun's to She/Her, clicked the button to display on the forum. When I logged in as instructor and viewed [this thread](http://localhost:1511/courses/s23/sample/forum/threads/7), the pronouns are indeed there. Nice!

However, when I went to [the same thread](http://localhost:1511/courses/s23/sample/forum/threads/7) as aphacker (or student), I received the following error:

```
Oh no! Something irrecoverable has happened...

app\exceptions\OutputException (Code: 0) thrown in /usr/local/submitty/site/app/libraries/Output.php (Line 372) by:
            throw new OutputException("{$e->getMessage()} in {$e->getFile()}:{$e->getLine()}");


Message:
Key "pronouns" does not exist as the array is empty. in /usr/local/submitty/site/app/templates/forum/CreatePost.twig:97

Stack Trace:
#0 /usr/local/submitty/site/app/views/forum/ForumThreadView.php(290): app\libraries\Output->renderTwigTemplate()
#1 unknown file(unknown line): app\views\forum\ForumThreadView->showForumThreads()
#2 /usr/local/submitty/site/app/libraries/Output.php(234): call_user_func_array()
#3 /usr/local/submitty/site/app/libraries/Output.php(203): app\libraries\Output->renderTemplate()
#4 /usr/local/submitty/site/app/controllers/forum/ForumController.php(1059): app\libraries\Output->renderOutput()
#5 unknown file(unknown line): app\controllers\forum\ForumController->showThreads()
#6 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(262): call_user_func_array()
#7 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(240): app\libraries\routers\WebRouter->run()
#8 /usr/local/submitty/site/public/index.php(131): app\libraries\routers\WebRouter::getWebResponse()
```

Could this error be addressed?

**Review by NiaHeermance — APPROVED** [review]
Works fully now! I approve.

**Review by NiaHeermance — CHANGES_REQUESTED** [review]
Actually, I just tested rolling back the migration and I get a crash. Re-migrating it does not fix it either. Here is the error:

```
Oh no! Something irrecoverable has happened...

app\exceptions\ConfigException (Code: 0) thrown in /usr/local/submitty/site/app/models/Config.php (Line 562) by:
                throw new ConfigException("Missing config setting '{$section}.{$key}' in configuration json file");


Message:
Missing config setting 'course_details.grade_inquiry_message' in configuration json file

Stack Trace:
#0 /usr/local/submitty/site/app/models/Config.php(507): app\models\Config->setConfigValues()
#1 /usr/local/submitty/site/app/libraries/Core.php(142): app\models\Config->loadCourseJson()
#2 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(278): app\libraries\Core->loadCourseConfig()
#3 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(183): app\libraries\routers\WebRouter->loadCourse()
#4 /usr/local/submitty/site/public/index.php(131): app\libraries\routers\WebRouter::getWebResponse()
```

Could this be fixed?

**Review by mahi-pas — CHANGES_REQUESTED** [review]
image
In my profile there should also be a field "Display Pronouns in Discussion Forum" and editing it would go to the pronouns modal, similar to first and last name opening the same modal

**Review by mahi-pas — CHANGES_REQUESTED** [review]
image
Is there a reason why there is a space between the parenthesis and the pronouns?

**Review by mahi-pas — COMMENTED** [review]
Functionality and code seem to be good. I don't have the expertise to verify the migrations.

File: migration/migrator/migrations/course/20230609110652_display_pronouns.py
Comment (williamjallen) [inline migration/migrator/migrations/course/20230609110652_display_pronouns.py:37]: We generally don't want to delete any data in the down migration in case something gets rolled back temporarily and then rolled forward again shortly thereafter.

File: migration/migrator/migrations/master/20230609110630_display_pronouns.py
Comment (williamjallen) [inline migration/migrator/migrations/master/20230609110630_display_pronouns.py:13]: this line isn't necessary

File: migration/migrator/migrations/course/20230609110652_display_pronouns.py
Comment (williamjallen) [inline migration/migrator/migrations/course/20230609110652_display_pronouns.py:20]: You don't need `pass` if you already have text in a given block of code in Python.

File: migration/migrator/migrations/course/20230609110652_display_pronouns.py
Comment (williamjallen) [inline migration/migrator/migrations/course/20230609110652_display_pronouns.py:38]: Same comment here

File: migration/migrator/migrations/master/20230609110630_display_pronouns.py
Comment (williamjallen) [inline migration/migrator/migrations/master/20230609110630_display_pronouns.py:91]: same comment here

File: migration/migrator/migrations/master/20230609110630_display_pronouns.py
Comment (williamjallen) [inline migration/migrator/migrations/master/20230609110630_display_pronouns.py:177]: same comment here

**Review by williamjallen — CHANGES_REQUESTED** [review]

**Review by NiaHeermance — CHANGES_REQUESTED** [review]
Migration rollback works now! There's still two things I notice:

- On the My Profile page, when you set the Display Pronouns in Discussion Forum option, the popup does not close automatically. This may be an intentional choice. However, some popups on that page do close automatically, so I want to make sure this was intentional.

- After changing whether you want your pronouns to be displayed, the text says "true" instead of "True" and "false" instead of "False." On a refresh, the capitalization is fixed, but the capitalization should probably be right to begin with.

File: site/app/templates/EditPronounsForm.twig
Comment (mahi-pas) [inline site/app/templates/EditPronounsForm.twig:32]: I'm pretty sure we shouldn't use `<br>` but rather use CSS. Either way, the gap is too large so it looks a bit odd
image

The gap should only be a little bit. But honestly, if you think it's unnecessary then we can just do without the gap and leave it to how it was before

**Review by mahi-pas — CHANGES_REQUESTED** [review]
Overall pretty good, I think I'll approve after this change

**Comment by NiaHeermance** [comment]
I now see this on the login page: 
image

**Comment by NiaHeermance** [comment]
I'm experiencing this now when I go to this page: http://localhost:1511/courses/f23/sample.

image

**Comment by yanliw123** [comment]
> I'm experiencing this now when I go to this page: http://localhost:1511/courses/f23/sample.
> 
> image

hmm, that's weird, I just tested the instructor and student account and both works fine without any errors when go to f23 sample course page

**Review by NiaHeermance — APPROVED** [review]
Everything works perfect. I approve!

**Comment by NiaHeermance** [comment]
I realized that I should be using submity_install when I first load a branch. I believe that was what was causing the echo on the side. Apologies for not realizing that. I have approved the PR.

File: site/app/controllers/UserProfileController.php
Comment (bmcutler) [inline site/app/controllers/UserProfileController.php:120]: what is the purpose of this 'before' field vs the 'display_pronouns' field?

File: site/app/controllers/admin/UsersController.php
Comment (bmcutler) [inline site/app/controllers/admin/UsersController.php:357]: I don't think we should be 'trim'ming a bool type value

File: site/app/views/forum/ForumThreadView.php
Comment (bmcutler) [inline site/app/views/forum/ForumThreadView.php:107]: why trim this?

File: site/app/views/forum/ForumThreadView.php
Comment (bmcutler) [inline site/app/views/forum/ForumThreadView.php:897]: why trim this?

File: site/app/views/forum/ForumThreadView.php
Comment (bmcutler) [inline site/app/views/forum/ForumThreadView.php:1007]: why trim this

File: site/app/views/forum/ForumThreadView.php
Comment (bmcutler) [inline site/app/views/forum/ForumThreadView.php:1133]: why trim this?

**Review by bmcutler — CHANGES_REQUESTED** [review]
asking questions, not sure if they are required changes

File: site/app/controllers/UserProfileController.php
Comment (yanliw123) [reply site/app/controllers/UserProfileController.php:120]: that was for error-checking purpose, I will remove the before field and only leave the 'pronouns' and 'display_pronouns’ fields for future use.

Comment (bmcutler) [reply site/app/controllers/UserProfileController.php:120]: please remove the before field ASAP.  thanks

**Review by bmcutler — CHANGES_REQUESTED** [review]
Unfortunately we have a bug with anonymous posts in the forum.
The pronouns are still displayed:

Screenshot 2023-07-26 at 3 02 20 PM

**Review by mahi-pas — CHANGES_REQUESTED** [review]
If you set pronouns in discussion forum to true
image
image
And then reload the page, it reverts to false.
image

 It appears the value works because it is shown in the discussion forum, but the display is wrong

**Review by mahi-pas — CHANGES_REQUESTED** [review]
image
When i make a thread on instructor and look at the student page to see if it's synced, the name and pronouns are not visible. if I refresh the page they are

**Review by NiaHeermance — COMMENTED** [review]
Why is display_pronouns added both to the course users table and the master user table? Shouldn't it only be added to the master user table?

**Comment by bmcutler** [comment]
The original design of the databases is that the course database is sufficient to render any page within the course.
One advantage of this is that an instructor can request a dump of the complete course database for their course and in theory import it to their own local instance of Submitty for archiving, etc.

The master database was created so that we knew which courses a specific student had access to.   (So we didn't have to loop over all courses on the system to check).

Much of the data about each user is therefore duplicated between the master and course databases.  That is a known redundancy/duplication (a slight but accepted disadvantage).  The trigger functions are supposed to ensure the duplication is in sync across all courses.

**Comment by NiaHeermance** [comment]
I see! Thanks for explaining that to me.

**Comment by bmcutler** [comment]
Also, in the original design, in theory, we should be able to destroy and then recreate the master database just with the info in each course database.  (Not sure if this is still true, and I've never attempted to do this.)

**Review by bmcutler — APPROVED** [review]
thanks for updating this PR :)

**Pull request merged**
